rpc_util.go

```diff
@@ -661,9 +681,10 @@ func recvAndDecompress(p *parser, s *transport.Stream, dc Decompressor, maxRecei
 			if err != nil {
 				return nil, status.Errorf(codes.Internal, "grpc: failed to decompress the received message %v", err)
 			}
+			expectedSize := int64(len(d) * compressionFactor)
 			// Read from LimitReader with limit max+1. So if the underlying
 			// reader is over limit, the result will be bigger than max.
-			d, err = ioutil.ReadAll(io.LimitReader(dcReader, int64(maxReceiveMessageSize)+1))
+			d, err = readAll(io.LimitReader(dcReader, int64(maxReceiveMessageSize)+1), expectedSize)
```

Comment:
If `expectedSize > maxReceiveMessageSize`, then there would be no reason to allocate a buffer that large.